### PR TITLE
feat(fsdp_tp): consumed-token metrics + global_batch_size in config

### PR DESCRIPTION
**Pull request by saforem2**
## Summary

Two observability features for `ezpz.examples.fsdp_tp`, so token-based loss
curves and the effective global batch size are visible in the run config / W&B.

### 1. Consumed-token metrics
Per-step and cumulative token accounting in the training loop:

| Metric | Meaning |
|--------|---------|
| `train/tokens` | Global tokens processed **this step** |
| `train/tokens_seen` | Cumulative tokens over the run (standard x-axis for loss-vs-tokens) |
| `train/tps` | Global tokens/sec (over distinct-data ranks) |
| `train/tps_per_gpu` | Per-rank tokens/sec (unchanged) |

**Global tokens/step = `batch × inp.shape[1] × dpsize`** — the **full,
pre-shard** sequence length (`inp.shape[1]`) times the data-parallel degree
`dpsize = dp_replicate × dp_shard`.

- `inp` is `Replicate()` across the tp group (the TP plan shards only the
  embedding *output*, never the input), so `inp.shape[1]` is the full sequence
  and is identical on every rank — the metric is exact and rank-invariant.
- **`tp` does not enter.** Under SP the tp ranks hold the *same* sequence
  (full-length logits under the default `Replicate()` output, or `Shard(1)`
  slices summing to the full length under fused-linear/loss-parallel), never
  distinct sequences. On the HF path (`tp` forced to 1, no SP) the tp-dim ranks
  see *duplicate* samples, so `dpsize` (not `world_size`) counts each distinct
  token once.

> Earlier iterations of this PR used `local_seq_len × world_size` (and briefly
> `local_seq_len × dpsize × tp`); both over-counted — by a full `~tp×` in the
> default eager path, or by `dpsize × (tp − remainder)` off-divisor in the
> seq-sharded paths. The final formula above is exact on every config.

### 2. `global_batch_size` in config
Computed in `train()` (not `parse_args`, because `--dp-shard -1` only resolves
once `WORLD_SIZE` is known), backfilled onto `args` so `History`'s `vars(args)`
and `wandb.config` both pick it up:

```python
args.global_batch_size = args.batch_size * dp_replicate * dp_shard
```

This is the general Megatron/NeMo formula `micro_batch × world_size ×
grad_accum / (tensor_parallel × pipeline_parallel)` specialized to this
example (no pipeline parallelism, no gradient accumulation → both factors 1;
`world_size / tp == dp_replicate × dp_shard`). It excludes `tp` (tp ranks share
the same samples).

## Tests
- `TestGlobalBatchSize` — `dp_replicate × dp_shard` formula; tp does **not** scale it; explicit `--dp-shard` path.
- `TestConsumedTokensAccounting` — full-length formula across SP-divisible, SP-remainder, HF-forced-tp1, and tp=1 paths.
- Full suite: `21 passed`; ruff clean.

## Docs
- New **Token Accounting** section in `docs/examples/fsdp-tp.md`.
- Re-anchored the `--8<--` snippet ranges that drifted as the source grew.

## Validation (2-node Sunspot, XPU)
ezpz installed from this PR head; world_size=24, dp_shard=12, dpsize=12, tp=2:
- `seq_len=512` → `train/tokens = 6144` (= 1×512×12), `tokens_seen` exact over 2666 steps.
- `seq_len=513` (odd) → `train/tokens = 6156` (= 1×513×12) — uses the true pre-shard length, not the padded/sharded one.
- `global_batch_size=12` present in the live W&B config (nested + flattened).
- Old formula would have reported 12288 / 12312 — confirming the fix.

### CODE REVIEW

File: src/ezpz/examples/fsdp_tp.py [hunk 1]
```diff
@@ -2947,10 +2978,20 @@ def train(
             # _slice_for_sequence_parallel), so the TP ranks process distinct
             # tokens too. Using dp_size would undercount global tps by ~tp.
             # At tp=1, world_size == dp_size so this is unchanged.
+            tokens_per_rank = args.batch_size * local_seq_len
+            # Global tokens processed THIS step across all ranks. world_size
+            # (not dp_size): under SP each tp rank holds a distinct seq shard,
+            # so tp ranks consume distinct tokens (at tp=1 these are equal).
+            tokens_this_step = tokens_per_rank * world_size
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Good catch — confirmed real. Fixed in 5af809b: the multiplier is now `dpsize * args.tp` (effective, post-force tp), which counts ranks holding *distinct* tokens. On the SP path (ezpz Transformer, tp>1) this equals `world_size` (no change); on the HF path (tp forced to 1, no SP) it collapses to `dpsize`, so `train/tokens{,_seen}`/`train/tps` no longer overcount by the requested TP factor. `TestConsumedTokensAccounting` now locks all three paths incl. the HF forced-tp=1 case.

File: src/ezpz/examples/fsdp_tp.py [hunk 2]
```diff
@@ -2255,6 +2276,10 @@ def train(
         from dataclasses import asdict
 
         wandb.config.update(asdict(config))  # type:ignore
+        # Also surface the resolved CLI args (incl. the backfilled
+        # global_batch_size) in the run config. allow_val_change since some
+        # keys may already be present from main()'s init.
+        wandb.config.update(vars(args), allow_val_change=True)  # type:ignore
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Confirmed and fixed in 98f277b. `args.tp` (HF force) and `args.loss_impl` (normalized to "compiled") are both mutated after the original push, so the run config recorded requested-not-effective settings. Split the update: `asdict(config)` stays early (ModelArgs is not mutated), but `vars(args)` now fires just before the training loop — after all args mutations and the `global_batch_size` backfill.

File: tests/test_fsdp_tp_dp_degrees.py [hunk 3]
```diff
@@ -114,3 +114,89 @@ def test_flags_are_ints(self):
...
+class TestGlobalBatchSize:
+    """global_batch_size = local_batch * dp_replicate * dp_shard.
+
+    Data parallelism spans BOTH the HSDP replicate dim and the FSDP shard
+    dim (the DistributedSampler shards over num_replicas = dp_replicate *
+    dp_shard). TP is NOT data parallel — tp ranks share the same batch — so
+    it must NOT scale the global batch. train() backfills
+    args.global_batch_size = args.batch_size * (dp_replicate * dp_shard);
+    these tests lock that formula via the resolver it's built on.
+    """
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_fsdp_tp_dp_degrees.py:128]: Added in 5af809b: `TestGlobalBatchSize.test_explicit_dp_shard` covers a concrete `--dp-shard 4` (world_size=8, tp=2, replicate=1) and asserts `gbs == batch * dp_replicate * dp_shard` with the explicit degree honored and tp excluded.

File: src/ezpz/examples/fsdp_tp.py [hunk 4]
```diff
@@ -2942,15 +2987,30 @@ def train(
...
+            #     dpsize was computed with the ORIGINAL tp, so the tp-dim ranks
+            #     see DUPLICATE samples. Here args.tp is the post-force value (1),
+            #     so dpsize * args.tp == dpsize and we don't overcount by ~tp.
+            # Using world_size unconditionally would overstate HF-path tokens by
+            # the requested TP factor.
+            tokens_per_rank = args.batch_size * local_seq_len
+            # Global tokens processed THIS step across all distinct-data ranks.
+            tokens_this_step = tokens_per_rank * dpsize * args.tp
+            tokens_seen += tokens_this_step
+            # Cumulative consumed training tokens — the standard x-axis for
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:3023]: Confirmed real and fixed in d826158. You were exactly right: rank 0 holds the ceil shard `ceil(seq/tp)`, so scaling it by `tp` overcounted the full sequence by `(tp - remainder)` per step — and since `inp = x[:, :-1]` the length is odd, so this was hitting most steps (and compounding in `tokens_seen`).

Fix computes from the full pre-shard length instead:
```python
tokens_this_step = args.batch_size * inp.shape[1] * dpsize
```
`inp` is `Replicate()` across the tp group (the TP plan shards only the embedding *output*), so `inp.shape[1]` is the full sequence and is identical on every rank — exact and rank-invariant. `tp` drops out (SP ranks hold shards of the *same* sequence; their lengths sum to the full length), and `dpsize` (not `world_size`) still handles the HF duplicate-data path. Verified exact across SP-divisible / SP-remainder / HF-forced-tp1 / tp=1; `TestConsumedTokensAccounting` now pins the off-divisor overcount it replaces.

File: tests/test_fsdp_tp_dp_degrees.py [hunk 5]
```diff
@@ -114,3 +114,164 @@ def test_flags_are_ints(self):
...
+    def _global_tokens(self, m, *, batch, seq_len, world_size, tp,
+                       dp_replicate, dp_shard):
+        """Global tokens/step exactly as train() computes it:
+        batch * seq_len * dpsize, where seq_len is the full (pre-shard)
+        length and dpsize is resolved from the ORIGINAL tp."""
+        rep, shard = m._resolve_dp_degrees(
+            world_size=world_size, tp=tp, dp_replicate=dp_replicate,
+            dp_shard=dp_shard,
+        )
+        return batch * seq_len * (rep * shard)
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_fsdp_tp_dp_degrees.py]: Fixed in 2a4f8f1: aligned the `_global_tokens` signature (hanging indent) and the `_resolve_dp_degrees` call (one kwarg per line) with the sibling `_gbs` helper. Note the whole-file `ruff format` diff is pre-existing (this module predates `ruff format` enforcement; CI runs `ruff check` only, which passes) — I kept the change scoped to the new helper rather than reformatting unrelated tests.

File: src/ezpz/examples/fsdp_tp.py [hunk 6]
```diff
@@ -2936,21 +2980,44 @@ def train(
...
+            #     is this rank's shard; scaling it by tp over-counts by
+            #     dpsize*(tp - remainder) when the sequence isn't divisible by tp
+            #     (and inp = x[:, :-1] makes an odd length common).
+            # On the HF path (tp forced to 1, no SP) the tp-dim ranks see
+            # DUPLICATE samples; multiplying by dpsize (not world_size) counts
+            # each distinct token exactly once.
+            tokens_per_rank = args.batch_size * local_seq_len
+            # Global tokens processed THIS step across all distinct-data ranks.
+            tokens_this_step = args.batch_size * inp.shape[1] * dpsize
+            tokens_seen += tokens_this_step
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:3022]: Updated the PR description to the corrected semantics: global tokens/step = `batch × inp.shape[1] × dpsize` (full pre-shard length, tp excluded), with a note on why the earlier `local_seq_len × world_size` / `× dpsize × tp` framings over-counted.

File: src/ezpz/examples/fsdp_tp.py [hunk 7]
```diff
@@ -2936,21 +2980,44 @@ def train(
...
-            # local_seq_len.
+            # Throughput.
             #   - train/tps_per_gpu : per-rank tokens/sec (torchtitan's `tgs`)
             #   - train/tps         : global tokens/sec across ALL ranks
-            # Multiply by world_size, not dp_size: under sequence parallelism
-            # each TP rank holds a DISTINCT seq shard (labels are sliced via
-            # _slice_for_sequence_parallel), so the TP ranks process distinct
-            # tokens too. Using dp_size would undercount global tps by ~tp.
-            # At tp=1, world_size == dp_size so this is unchanged.
+            # tokens_per_rank uses `local_seq_len` (= pred.shape[1]) because it
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:2996]: Fixed in 2a4f8f1: reworded the `train/tps` comment from "global tokens/sec across ALL ranks" to "global tokens/sec (over distinct-data ranks; see the dpsize rationale below)" so it no longer implies world_size scaling for the HF forced-tp1 duplicate-rank case.

File: docs/examples/fsdp-tp.md [hunk 8]
```diff
@@ -549,6 +549,49 @@ metrics["train/mfu"] = compute_mfu(_model_flops, t2 - t0)
...
+```
+
+Per step, the example reports how many tokens were consumed:
+
+| Metric | Meaning |
+|--------|---------|
+| `train/tokens` | Global tokens processed **this step** |
+| `train/tokens_seen` | Cumulative tokens over the whole run (the standard x-axis for loss-vs-tokens curves) |
+| `train/tps` | Global tokens/sec across all ranks |
+| `train/tps_per_gpu` | Per-rank tokens/sec |
```
[the remark this answers is not in the document]
Comment (saforem2) [reply docs/examples/fsdp-tp.md:578]: Fixed in 2a4f8f1: the docs table row now reads "Global tokens/sec (over distinct-data ranks — see the `dp_size` note below)", consistent with the explanation and the implementation.
